src/client.rs

```diff
@@ -62,11 +63,64 @@ impl CoinGeckoClient {
     /// let client = CoinGeckoClient::new("https://some.url");
     /// ```
     pub fn new(host: &'static str) -> Self {
-        CoinGeckoClient { host }
+        CoinGeckoClient {
+            host,
+            api_key: None,
+        }
+    }
+
+    /// Creates a new CoinGeckoClient client with a custom host url and API key
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use coingecko::CoinGeckoClient;
+    /// let client = CoinGeckoClient::new_with_api_key("https://some.url", "123456789");
+    /// ```
+    pub fn new_with_api_key(host: &'static str, api_key: &'static str) -> Self {
+        CoinGeckoClient {
+            host,
+            api_key: Some(api_key),
+        }
     }
 
-    async fn get<R: DeserializeOwned>(&self, endpoint: &str) -> Result<R, Error> {
-        reqwest::get(format!("{host}/{ep}", host = self.host, ep = endpoint))
+    /// Gets a URL for the provided endpoint and optional params.
+    /// The endpoint must be prefixed with a /.
+    /// If params are specified, they must be prefixed with a ?.
+    /// If an API key is present, the x_cg_pro_api_key param is added.
+    pub fn get_url(&self, endpoint: &str, params: Option<&str>) -> String {
```

Comment:
If we are keeping this function it would make sense for `params` to be `params: impl Iter<Item=(&str, &str)>` and just do all the formatting here, it would avoid the extra logic for how to merge or change params since you would always just have a list of params.